src/ucs/arch/cpu.c

```diff
@@ -0,0 +1,106 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2019. ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#include <ucs/arch/cpu.h>
+#include <ucs/arch/generic/cpu.h>
+#include <ucs/sys/sys.h>
+#include <ucs/sys/string.h>
+#include <ucs/sys/stubs.h>
+#include <ucs/type/init_once.h>
+
+#define UCS_CPU_CACHE_FILE_FMT     "/sys/devices/system/cpu/cpu%d/cache/index%d/%s"
+#define UCS_CPU_CACHE_LEVEL        "level"
+#define UCS_CPU_CACHE_TYPE         "type"
+#define UCS_CPU_CACHE_SIZE         "size"
+
+/* cache size array. index - cache type (ucs_cpu_cache_type_t), value - cache value,
+ * 0 means cache is not supported */
+static size_t ucs_cpu_cache_size[UCS_CPU_CACHE_LAST] = {0};
+
+static ucs_init_once_t ucs_cache_read_once = UCS_INIT_ONCE_INIITIALIZER;
+
+/* cache datatypes */
+struct { /* cache info obtained from [linux] kernel */
+    int             level;
+    const char     *type;
+} const ucs_cpu_cache_sysfs_name[] = {
+    [UCS_CPU_CACHE_L1d] = {.level = 1, .type = "Data"},
+    [UCS_CPU_CACHE_L1i] = {.level = 1, .type = "Instruction"},
+    [UCS_CPU_CACHE_L2]  = {.level = 2, .type = "Unified"},
+    [UCS_CPU_CACHE_L3]  = {.level = 3, .type = "Unified"}
+};
+
+
+static void ucs_sysfs_get_cache_size(size_t *cache, size_t length)
+{
+    char type_str[32];  /* Data/Instruction/Unified */
+    char size_str[32];  /* memunits */
+    int i;
+    int cpu;
+    long level;
+    ssize_t file_size;
+    size_t c;
+    ucs_status_t status;
+
+    ucs_assert(length <= ucs_static_array_size(ucs_cpu_cache_sysfs_name));
+
+    cpu = ucs_get_first_cpu();
+
+    for (i = 0;; i++) {
+        file_size = ucs_read_file_str(type_str, sizeof(type_str), 1, UCS_CPU_CACHE_FILE_FMT,
+                                      cpu, i, UCS_CPU_CACHE_TYPE);
+        if (file_size < 0) {
+            return; /* no more files */
+        }
+
+        status = ucs_read_file_number(&level, 1, UCS_CPU_CACHE_FILE_FMT,
+                                      cpu, i, UCS_CPU_CACHE_LEVEL);
+        if (status != UCS_OK) {
+            return; /* no more files */
+        }
+
+        /* ok, we found valid directory, let's try to read cache size */
+        file_size = ucs_read_file_str(size_str, sizeof(size_str), 1, UCS_CPU_CACHE_FILE_FMT,
+                                      cpu, i, UCS_CPU_CACHE_SIZE);
+        if (file_size < 0) {
+            return; /* no more files */
+        }
+
+        /* now lookup for cache size entry */
+        for (c = 0; c < length; c++) {
+            if ((ucs_cpu_cache_sysfs_name[c].level == level) &&
+                !strcasecmp(ucs_cpu_cache_sysfs_name[c].type, ucs_strtrim(type_str))) {
```

Comment:
can we do strtrim only once, after reading it?

Reply:
trimmed once after read complete